src/core/Carousel/Carousel.tsx

```diff
@@ -0,0 +1,494 @@
+        setCurrentIndex((old) => (slideCount + old + 1) % slideCount);
+        break;
+      }
+    }
+  };
+
+  return (
+    <section
+      className='iui-carousel'
+      aria-roledescription='carousel'
+      tabIndex={0}
+      onKeyDown={handleKeyDown}
+      onKeyUp={handleKeyUp}
+      {...rest}
+      id={id}
+    >
+      <CarouselContext.Provider
+        value={{
+          currentIndex,
+          setCurrentIndex,
+          slideCount,
+          setSlideCount,
+          keysPressed,
+          idPrefix: id,
+          isManuallyUpdating,
+        }}
+      >
+        {children}
+      </CarouselContext.Provider>
+    </section>
+  );
+};
+
+/**
+ * CarouselSlider is the scrollable list that should consist of CarouselSlide components.
+ */
+const CarouselSlider = React.forwardRef<
+  HTMLOListElement,
+  React.ComponentPropsWithoutRef<'ol'>
+>((props, ref) => {
+  const context = React.useContext(CarouselContext);
+
+  if (!context) {
+    throw new Error('CarouselSlider must be used within Carousel');
+  }
+
+  const { currentIndex, setSlideCount, idPrefix } = context;
+  const sliderRef = React.useRef<HTMLOListElement>(null);
+  const refs = useMergedRefs(sliderRef, ref);
+
+  React.useLayoutEffect(() => {
+    setSlideCount(React.Children.count(props.children));
+  }, [props.children, setSlideCount]);
+
+  const items = React.useMemo(
+    () =>
+      React.Children.map(props.children, (child, index) =>
+        React.isValidElement(child)
+          ? React.cloneElement(child, {
+              id: `${idPrefix}--slide-${index}`,
+              index,
+            })
+          : child,
+      ) ?? [],
+    [props.children, idPrefix],
+  ) as React.ReactNode[];
+
+  React.useEffect(() => {
+    if (!sliderRef.current) {
+      return;
+    }
+
+    sliderRef.current.children.item(currentIndex)?.scrollIntoView();
+  }, [currentIndex]);
+
+  return (
+    <ol
+      aria-live='polite'
+      {...props}
+      ref={refs}
+      className={cx('iui-carousel-slider', props.className)}
+    >
+      {items}
+    </ol>
+  );
+});
+
+/**
+ * CarouselSlide is used for the actual slide content. The content can be specified through `children`.
+ *
+ * It is recommended that the slide content bring its own dimensions (esp. height) and that
+ * the dimensions should be the same for all slides.
+ */
+const CarouselSlide = React.forwardRef<
+  HTMLLIElement,
+  { index?: number } & React.ComponentPropsWithoutRef<'li'>
+>(({ index, ...props }, ref) => {
+  const { className, children, ...rest } = props;
+
+  const context = React.useContext(CarouselContext);
+  if (!context) {
+    throw new Error('CarouselSlider must be used within Carousel');
+  }
+
+  const { currentIndex, isManuallyUpdating, setCurrentIndex } = context;
+
+  const updateActiveIndex = React.useCallback(() => {
+    if (!isManuallyUpdating.current) {
+      setCurrentIndex((prev) => (prev !== index ? index : prev));
```

Comment:
You still have `currentIndex` in deps list. So maybe just move this condition to the `if` condition above or have it in different `useEffect`.

Reply:
Done